ethcore/node-filter/src/lib.rs

```diff
@@ -37,28 +37,44 @@ extern crate tempdir;
 #[macro_use]
 extern crate log;
 
+use std::collections::{HashMap, VecDeque};
 use std::sync::Weak;
 
-use ethcore::client::{BlockChainClient, BlockId};
+use ethcore::client::{BlockChainClient, BlockId, ChainNotify, NewBlocks};
+
 use ethereum_types::{H256, Address};
 use ethabi::FunctionOutputDecoder;
 use network::{ConnectionFilter, ConnectionDirection};
 use devp2p::NodeId;
+use devp2p::MAX_NODES_IN_TABLE;
+use parking_lot::RwLock;
 
 use_contract!(peer_set, "res/peer_set.json");
 
 /// Connection filter that uses a contract to manage permissions.
 pub struct NodeFilter {
 	client: Weak<BlockChainClient>,
 	contract_address: Address,
+	cache: RwLock<Cache>
+}
+
+struct Cache {
+	cache: HashMap<NodeId, bool>,
+	order: VecDeque<NodeId>
 }
 
+pub const CACHE_SIZE: usize = MAX_NODES_IN_TABLE + 1024;
```

Comment:
Just curious, why +1024? :-)

Reply:
We are chaining reserved nodes with nodes from table here:

https://github.com/paritytech/parity-ethereum/blob/d89b8d904ff28382d3692290a936706304627255/util/network-devp2p/src/host.rs#L587-L591

So if the cache size would be the same as node table size, in case when node table is full, the `reserved_peers.len` amount of unnecessary calls will be done to the contract, since we hit limits of cache.

Reply:
Got it. But IMO it's also worth some explanation

Reply:
I agree! I would just add a comment there.